test/node/object_id_tests.js

```diff
@@ -24,6 +24,192 @@ describe('ObjectId', function () {
+    }
+    var objectBufferId = {
+      id: validBuffer1,
+      toHexString: newToHexString
+    };
+    var objectBufferFromArray = {
+      id: validBuffer2,
+      toHexString: newToHexString
+    };
+    expect(Buffer.from(new ObjectId(objectBufferId).id)).to.deep.equal(bufferToHex);
+    expect(Buffer.from(new ObjectId(objectBufferFromArray).id)).to.deep.equal(bufferToHex);
+  });
+
+  it('should throw error if object with non-Buffer non-string id is passed in', function () {
+    var objectNumId = {
+      id: 5
+    };
+    var objectNullId = {
+      id: null
+    };
+    expect(() => new ObjectId(objectNumId)).to.throw(TypeError);
+    expect(() => new ObjectId(objectNullId)).to.throw(TypeError);
+  });
+
+  it('should throw an error if object with invalid string id is passed in', function () {
+    var objectInvalidString = {
+      id: 'FFFFFFFFFFFFFFFFFFFFFFFG'
+    };
+    expect(() => new ObjectId(objectInvalidString)).to.throw(TypeError);
+  });
+
+  it('should correctly create ObjectId from object with invalid string id and toHexString method', function () {
+    function newToHexString() {
+      return 'BBBBBBBBBBBBBBBBBBBBBBBB';
+    }
+    var objectInvalidString = {
+      id: 'FFFFFFFFFFFFFFFFFFFFFFFG',
+      toHexString: newToHexString
+    };
+    var buf = Buffer.from('BBBBBBBBBBBBBBBBBBBBBBBB', 'hex');
+    expect(Buffer.from(new ObjectId(objectInvalidString).id)).to.deep.equal(buf);
+  });
+
+  it('should throw an error if object with invalid Buffer id is passed in', function () {
+    var objectInvalidBuffer = {
+      id: Buffer.from([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13])
+    };
+    expect(() => new ObjectId(objectInvalidBuffer)).to.throw(TypeError);
+  });
+
+  it('should correctly create ObjectId from object with invalid Buffer id and toHexString method', function () {
+    function newToHexString() {
+      return 'BBBBBBBBBBBBBBBBBBBBBBBB';
+    }
+    var objectInvalidBuffer = {
+      id: Buffer.from([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]),
+      toHexString: newToHexString
+    };
+    var buf = Buffer.from('BBBBBBBBBBBBBBBBBBBBBBBB', 'hex');
+    expect(Buffer.from(new ObjectId(objectInvalidBuffer).id)).to.deep.equal(buf);
+  });
+
+  const numericIO = [
+    { input: 42, output: 42 },
+    { input: 0x2a, output: 0x2a },
+    { input: 4.2, output: 4 },
+    { input: NaN, output: 0 }
+  ];
+
+  for (const { input, output } of numericIO) {
+    it(`should correctly create ObjectId from ${input} and result in ${output}`, function () {
+      const objId = new ObjectId(input);
+      expect(objId).to.have.property('id');
+      expect(objId.id).to.be.instanceOf(Buffer);
+      expect(objId.id.readUInt32BE(0)).to.equal(output);
+    });
+  }
+
+  it('should correctly create ObjectId undefined or null', function () {
+    const objNull = new ObjectId(null);
+    const objNoArg = new ObjectId();
+    const objUndef = new ObjectId(undefined);
+    expect(objNull.id).to.be.instanceOf(Buffer);
```

Comment:
can we also put the expected buffer value check for these test cases?

Reply:
We can't check for an expected buffer value for these cases, since each time this runs it'll generate a random Buffer based on the timestamp.